asabiris/orchestration/webhandler.py

```diff
@@ -0,0 +1,167 @@
+import logging
+
+import asab.web.rest
+
+import aiohttp.web
+import aiohttp.payload_streamer
+
+import jinja2
+
+from .emailschema import email_schema
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class WebHandler(object):
+
+	def __init__(self, app):
+		self.App = app
+
+		web_app = app.WebContainer.WebApp
+		web_app.router.add_put(r"/send_mail", self.send_mail)
+		web_app.router.add_put(r"/render", self.render)
+
+
+	@asab.web.rest.json_schema_handler(email_schema)
+	async def send_mail(self, request, *, json_data):
+		"""
+		This endpoint is for sending emails.
+		```
+		1) It collects the basic email info (to, cc, bcc, subject, from)
+		2) It renders the email body based on the template
+		3) Optionally it adds attachments:
+
+			3.1) The attachment is renders by this service.
+
+			3.2) The attachment is provided by the caller.
+
+		```
+		Example body:
+
+		```
+		{
+			"to": ['Tony.Montana@Goodfellas.com'],
+			"cc": ['Jimmy2.times@Goodfellas.com'],
+			"bcc": ['Henry.Hill@Goodfellas.com'],
+			"subject": "Lufthansa Hiest",
+			"from": "Jimmy.Conway@Goodfellas.com",
+			"body": {
+				"template": "test.md",
+				"params": {
+					"Name": "Toddy Siciro"
+			}
+		},
+		"attachments": [
+			{
+			"template": "test.md",
+			"params": {
+				"Name": "Michael Corleone"
+				},
+			"format": "pdf",
+			"filename": "Made.pdf"
+			}]
+		}
+
+		```
+		Attached will be retrieved from request.conent when rendering the email is not required.
+
+		Example of the email body template:
+		```
+		SUBJECT: Automated email for {{name}}
+
+		Hi {{name}},
+
+		this is a nice template for an email.
+		It is {{time}} to leave.
+
+		Br,
+		Your automated ASAB report
+		```
+
+		It is a markdown template.
+		---
+		tags: ['Send mail']
+		"""
+
+		try:
+			await self.App.SendMailOrchestrator.send_mail(
+				email_to=json_data["to"],
+				body_template=json_data["body"]["template"],
+				email_cc=json_data.get("cc", []),  # Optional
+				email_bcc=json_data.get("bcc", []),  # Optional
+				email_subject=json_data.get("subject", None),  # Optional
+				email_from=json_data.get("from"),
+				body_params=json_data["body"].get("params", {}),  # Optional
+				attachments=json_data.get("attachments", []),  # Optional
+			)
+
+		except KeyError as e:
+			raise aiohttp.web.HTTPNotFound(text="{}".format(e))
+
+		except jinja2.exceptions.UndefinedError as e:
+			raise aiohttp.web.HTTPBadRequest(text="Jinja2 error: {}".format(e))
+		
+		# More specific exception handling goes here so that the service provides nice output
+
+		return asab.web.rest.json_response(request, {"result": "OK"})
+
+
+	# TODO: JSON schema
```

Comment:
@ateska What are we actually validating here?

Reply:
The body of the PUT request of the `/render` method.